File: eigenvectors/eigenvectors.m
function eigenvectors()
    format long;
    A = [-10, 4; -4, 0];
    x0 = [-3; 2];
    tolerance = 1e-4;
    max_iterations = 50;
    lambda = power_iteration(A, x0, tolerance, max_iterations);
end

function [lambda, x] = power_iteration(A, x0, tolerance, max_iterations)
    x = x0 / norm(x0, inf);
    [~, index] = max(abs(x));
    if x(index) < 0
        x = -x;
    endif

    lambda = 0;
    for i = 1:max_iterations
        A_new = A * x;
        norm_new = norm(A_new, inf);
        if norm_new == 0
            lambda = 0;
            x = zeros(size(x));
            return;
        endif

        x_new = A_new / norm_new;
        [~, index] = max(abs(x_new));
        if x_new(index) < 0
            x_new = -x_new;
        endif

        lambda_new = (x' * A_new) / (x' * x);
        if norm(x_new - x, inf) < tolerance
            lambda = lambda_new
            x = x_new
            return;
        endif

        x = x_new;
        lambda = lambda_new;
    endfor
end

File: eigenvectors/eigenvectors.py
def power_iteration(A: list[list[float]], x0: list[float], tolerance: float, max_iterations: int = 50):
    x = [float(value) for value in x0]
    if not x:
        return 0.0, []

    scale = max(abs(value) for value in x)
    if scale == 0:
        return 0.0, [0.0] * len(x)

    x = [value / scale for value in x]
    dominant_index = max(range(len(x)), key=lambda index: abs(x[index]))
    if x[dominant_index] < 0:
        x = [-value for value in x]

    eigenvalue = 0.0
    for _ in range(max_iterations):
        x_new_raw = [sum(A[i][j] * x[j] for j in range(len(A))) for i in range(len(A))]
        norm = max(abs(value) for value in x_new_raw)
        if norm == 0:
            return 0.0, [0.0] * len(x)

        x_new = [value / norm for value in x_new_raw]
        dominant_index = max(range(len(x_new)), key=lambda index: abs(x_new[index]))
        if x_new[dominant_index] < 0:
            x_new = [-value for value in x_new]

        denominator = sum(value * value for value in x)
        if denominator == 0:
            return 0.0, [0.0] * len(x)

        eigenvalue = sum(x[i] * x_new_raw[i] for i in range(len(x))) / denominator
        err = max(abs(x_new[i] - x[i]) for i in range(len(x)))
        x = x_new
        if err < tolerance:
            return eigenvalue, x

    return eigenvalue, x